app/shared_channel_notifier.go

```diff
@@ -4,51 +4,140 @@
 
 import (
+	"fmt"
+
+	"github.com/pkg/errors"
+
 	"github.com/mattermost/mattermost-server/v5/mlog"
 	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/mattermost/mattermost-server/v5/services/sharedchannel"
 )
 
-var SharedChannelEventsForSync model.StringArray = []string{
+var sharedChannelEventsForSync model.StringArray = []string{
 	model.WEBSOCKET_EVENT_POSTED,
 	model.WEBSOCKET_EVENT_POST_EDITED,
 	model.WEBSOCKET_EVENT_POST_DELETED,
 	model.WEBSOCKET_EVENT_REACTION_ADDED,
 	model.WEBSOCKET_EVENT_REACTION_REMOVED,
 }
 
-// ServerSyncSharedChannelHandler is called when a websocket event is received by a cluster node.
-// Only on the leader node it will notify the sync service to start syncing content for the given
+var sharedChannelEventsForInvitation model.StringArray = []string{
+	model.WEBSOCKET_EVENT_DIRECT_ADDED,
+}
+
+// SharedChannelSyncHandler is called when a websocket event is received by a cluster node.
+// Only on the leader node it will notify the sync service to perform necessary updates to the remote for the given
 // shared channel.
-func (s *Server) ServerSyncSharedChannelHandler(event *model.WebSocketEvent) {
+func (s *Server) SharedChannelSyncHandler(event *model.WebSocketEvent) {
 	syncService := s.GetSharedChannelSyncService()
-	if !isEligibleForContentSync(syncService, event) {
-		return
+	if isEligibleForEvents(syncService, event, sharedChannelEventsForSync) {
+		err := handleContentSync(s, syncService, event)
+		if err != nil {
+			mlog.Warn(
+				err.Error(),
+				mlog.String("event", event.EventType()),
+				mlog.String("action", "content_sync"),
+			)
+		}
+	} else if isEligibleForEvents(syncService, event, sharedChannelEventsForInvitation) {
+		err := handleInvitation(s, syncService, event)
+		if err != nil {
+			mlog.Warn(
+				err.Error(),
+				mlog.String("event", event.EventType()),
+				mlog.String("action", "invitation"),
+			)
+		}
 	}
+}
 
-	channel, err := s.Store.Channel().Get(event.GetBroadcast().ChannelId, true)
+func isEligibleForEvents(syncService SharedChannelServiceIFace, event *model.WebSocketEvent, events model.StringArray) bool {
+	return syncServiceEnabled(syncService) &&
+		eventHasChannel(event) &&
+		events.Contains(event.EventType())
+}
+
+func eventHasChannel(event *model.WebSocketEvent) bool {
+	return event.GetBroadcast() != nil &&
+		event.GetBroadcast().ChannelId != ""
+}
+
+func syncServiceEnabled(syncService SharedChannelServiceIFace) bool {
+	return syncService != nil &&
+		syncService.Active()
+}
+
+func handleContentSync(s *Server, syncService SharedChannelServiceIFace, event *model.WebSocketEvent) error {
+	channel, err := findChannel(s, event.GetBroadcast().ChannelId)
 	if err != nil {
-		mlog.Warn(
-			"Received websocket message that is eligible for shared channel sync but channel does not exist",
-			mlog.String("event", event.EventType()),
-		)
-		return
+		return err
 	}
 
-	if !channel.IsShared() {
-		return
+	if channel != nil || channel.IsShared() {
```

Comment:
Shouldn't this be &&?  With || it will notify whether channel is shared or not.

Reply:
Totally, amended!